gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java

```
@@ -0,0 +1,175 @@
+    }
+
+    this.outputStream = this.fs.create(this.stagingFile, true);
+    this.prependSize = properties.getPropAsBoolean(ConfigurationKeys.SIMPLE_WRITER_PREPEND_SIZE, true);
+    this.recordsWritten = 0;
+    this.bytesWritten = 0;
+    this.closed = false;
+  }
+  /**
+   * Write a source record to the staging file
+   *
+   * @param record data record to write
+   * @throws java.io.IOException if there is anything wrong writing the record
+   */
+  @Override
+  public void write(byte[] record) throws IOException {
+    byte[] toWrite = record;
+    Preconditions.checkNotNull(record);
+    if (recordDelimiter != null) {
+      toWrite = Arrays.copyOf(record, record.length + 1);
+      toWrite[toWrite.length - 1] = recordDelimiter;
+    }
+    if (prependSize) {
+      Long recordSize = new Long(toWrite.length);
+      ByteBuffer buf = ByteBuffer.allocate(Long.SIZE / 8);
+      buf.putLong(recordSize);
+      toWrite = ArrayUtils.addAll(buf.array(), toWrite);
+    }
+    this.outputStream.write(toWrite);
+    bytesWritten += (toWrite.length);
+    recordsWritten++;
+  }
+
+  /**
+   * Commit the data written to the final output file.
+   *
+   * @throws java.io.IOException if there is anything wrong committing the output
+   */
+  @Override
+  public void commit() throws IOException {
+    this.close();
+    if (!this.fs.exists(this.stagingFile)) {
+      throw new IOException(String.format("File %s does not exist", this.stagingFile));
+    }
+
+    LOG.info(String.format("Moving data from %s to %s", this.stagingFile, this.outputFile));
+    // For the same reason as deleting the staging file if it already exists, deleting
+    // the output file if it already exists prevents task retry from being blocked.
+    if (this.fs.exists(this.outputFile)) {
+      LOG.warn(String.format("Task output file %s already exists", this.outputFile));
+      this.deletePath(this.outputFile, false);
+    }
+    this.fs.rename(this.stagingFile, this.outputFile);
```

Similar as `this.deletePath()`, it's better to have a wrapper around `fs.rename()` otherwise it fails silently.